synapse/storage/databases/main/receipts.py

```diff
@@ -941,34 +941,39 @@ async def _background_receipts_linearized_unique_index(
         receipts."""
 
         def _remote_duplicate_receipts_txn(txn: LoggingTransaction) -> None:
+            if isinstance(self.database_engine, PostgresEngine):
+                ROW_ID_NAME = "ctid"
+            else:
+                ROW_ID_NAME = "rowid"
+
             # Identify any duplicate receipts arising from
             # https://github.com/matrix-org/synapse/issues/14406.
             # We expect the following query to use the per-thread receipt index and take
             # less than a minute.
-            sql = """
-                SELECT MAX(stream_id), room_id, receipt_type, user_id
+            sql = f"""
+                SELECT MAX(stream_id), {ROW_ID_NAME}, room_id, receipt_type, user_id
                 FROM receipts_linearized
                 WHERE thread_id IS NULL
                 GROUP BY room_id, receipt_type, user_id
                 HAVING COUNT(*) > 1
             """
             txn.execute(sql)
-            duplicate_keys = cast(List[Tuple[int, str, str, str]], list(txn))
+            duplicate_keys = cast(List[Tuple[int, int, str, str, str]], list(txn))
 
             # Then remove duplicate receipts, keeping the one with the highest
-            # `stream_id`. There should only be a single receipt with any given
-            # `stream_id`.
-            for max_stream_id, room_id, receipt_type, user_id in duplicate_keys:
-                sql = """
+            # `stream_id`. Since there might be duplicate rows with the same
+            # `stream_id`, we delete by the rowid instead.
+            for _, row_id, room_id, receipt_type, user_id in duplicate_keys:
+                sql = f"""
                     DELETE FROM receipts_linearized
                     WHERE
                         room_id = ? AND
                         receipt_type = ? AND
                         user_id = ? AND
                         thread_id IS NULL AND
-                        stream_id < ?
+                        {ROW_ID_NAME} != ?
```

Comment:
Do we want to keep the `stream_id <` clause here in-case things have updated in the mean time?

Reply:
Could we end up deleting _all_ rows for a given (room_id, receipt_type, user_id) if we were to do that?
ie. is it possible for the row with the highest ctid not have the highest stream id?

Reply:
> Could we end up deleting _all_ rows for a given (room_id, receipt_type, user_id) if we were to do that? ie. is it possible for the row with the highest ctid not have the highest stream id?

I think `ctid` wraps so maybe -- I think if we used both (`stream_id <= ? AND ctid != ?`) I think we have potential to *still* end up with duplicates.

Reply:
We talked a bit more about this: https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$2S69hM_BtY08U1_2xR3t73aDVh7Vb82dBAY-yc2xyOM?via=matrix.org&via=element.io&via=beeper.com

The tl;dr we came to was:

* Why is `stream_id` not unique? 🤷 
* It might be better to try deleting everything and inserting the one "correct" row.

----

Unfortunately that is harder than it looks since the group-by behavior of SQLite and PostgreSQL is different (Postgres requires all columns to be grouped or aggregated; sqlite will just select the extra columns when you use max...)

Reply:
@erikjohnston and I talked about this, what we want to be able to do is:

* Find the max stream ID for each duplicate room, receipt type, user.
* Find the ctid of that row
* Delete everything for that room, receipt type, user which is not that ctid.

We could likely do the first two steps in one go using different logic on sqlite vs. postgres (distinct on vs. grouping), but I think it is easier to just do a second query to select the ctid inside the loop.